crates/core/transaction/src/plan/build.rs

```diff
@@ -14,72 +17,35 @@ use crate::{
 };
 
 impl TransactionPlan {
-    /// Build the transaction this plan describes.
-    ///
-    /// To turn a transaction plan into an unauthorized transaction, we need:
-    ///
-    /// - `fvk`, the [`FullViewingKey`] for the source funds;
-    /// - `witness_data`, the [`WitnessData`] used for proving;
-    ///
-    pub fn build(
+    /// Builds a [`TransactionPlan`] by slotting in the
+    /// provided prebuilt actions instead of using the
+    /// [`ActionPlan`]s in the TransactionPlan.
+    /// Arguments:
+    ///     self: `TransactionPlan`
+    ///     fvk: `FullViewingKey`
+    ///     witness_data: `WitnessData`
+    /// Returns: `Transaction`
+    pub fn build_unauth_with_actions(
         self,
-        fvk: &FullViewingKey,
+        mut actions: Vec<Action>,
```

Comment:
Shouldn't `actions` be immutable, since it's preconstructed data?

Reply:
Right now, I'd call the `actions` **_semi_** preconstructed data. If we think about this from the perspective of a web-worker, they each call [build_action](https://github.com/penumbra-zone/penumbra/blob/tx-decompose/crates/wasm/src/wasm_planner.rs#L76) which internally matches on the `Action` type. `build_action` should only be called for actions that are computationally intensive to compute. It therefore only precomputes the `output, spend, swap, swap-claim, and delegator-vote` actions. 

If we don't mark it as mutable, how should we handle pushing other [actions](https://github.com/penumbra-zone/penumbra/blob/tx-decompose/crates/core/transaction/src/plan/build.rs#L79) to the transaction body?

Reply:
We explicitly shouldn't handle that, we should require that all of the actions are built and passed in, rather than trying to manage things case by case.

As a first-pass implementation strategy, I think it would be a good idea to unconditionally spawn a new web worker for each ActionPlan, but even if we did something more sophisticated, it wouldn't change this interface.

In general we should try to avoid action-by-action special case handling and instead handle ActionPlans and Actions uniformly, as with some of the other PR suggestions.

Reply:
The web-workers are heavy instances (rather than simply being lightweight threads) since they replicate the underlying VM state. Spawning unnecessary web-workers would degrade performance since it increases the message passing overhead.

Reply:
There are also 22 variants for an `ActionPlan`, which means we'd need to spawn more threads than the number of physical cores on the device.

Reply:
I implemented your suggestion as a first-pass strategy. The pre-built actions are now _immutable_ by default. I'll see how the performance scales in terms of execution time and latency in the **webgpu** repository and report back.